pkg/apis/controlplane/v1beta2/types.go

```diff
@@ -61,6 +61,16 @@ type ExternalEntityReference struct {
 	Namespace string `json:"namespace,omitempty" protobuf:"bytes,2,opt,name=namespace"`
 }
 
+// EntityReference represents a reference to either a Pod or an ExternalEntity.
+// TODO: replace Pod and ExternalEntity in GroupMember to embed EntityReference
+//  when controlplane version is bumped?
+type EntityReference struct {
```

Comment:
I did not see this is used in the controlplane API? Is it for internal use only? Then no need to define it here?

Reply:
I remember `controlplane/v1beta2/types.go` should always be consistent with `controlplane/types.go` as the latest API version. Or do you mean that this should be placed in somewhere other than controlplane?

Reply:
I do not feel that is necessary, but maybe I missed some discussions before. @tnqn ?

Reply:
@antoninbas on this

Reply:
I think Jianjun's point is that this type should not really be part of the API (yet?). You do not use it in the controlplane API to exchange information between Controller & Agent: you just use it internally in the Agent. As such, it would make more sense for now to define the type somewhere under `pkg/agent` and not `pkg/apis/controlplane`.

Your comment holds when a type is actually part of the API:
> Similar to the versioned APIs, the definitions for the internal structs are in `pkg/apis/<group>/types.go`. Edit those files to reflect the change you want to make. Keep in mind that the internal structs must be able to express all of the versioned APIs.

Given that (based on your TODO) you intend to use this type in a *future* version of the API, I would recommend not adding it to `v1beta2.types.go`. One thing I can think of is defining the type in `controlplane/types.go` only for now (and not in `v1beta2/types.go`), and using `controlplane.EntityReference` wherever you use `v1beta2.EntityReference` now. I have to confess I don't know why we use the versioned type `v1beta2.PodReference` in the Agent currently, instead of the internal type (`controlplane.PodReference`). Maybe @tnqn knows the reason.

Reply:
`/apis/controlplane` is for API communication. Some files/code seem redundant but they are necessary compatibility consideration. For example, a struct in `controlplane/types.go` must also be in `controlplane//v1beta1/types.go` so they can be converted to each other.

`controlplane/types.go` is supposed to be used by the api provider itself only to convert different versions of versioned APIs like v1beta1 and v1beta2. So "internal" means antrea-controller internal as no other components including antrea-agent cannot actually receive these types from any API.

The reason I used `v1beta1.PodReference` was only to reuse an existing available struct when I wanted to put Pod Namespace and Name into it. But the data is produced and consumed agent internal only, so it can just use its own struct or reuse some more generic one like `apimachinerytypes.NamespacedName` if we don't care the entity type. But it shouldn't change `apis/controlplane` for this purpose as it expands API definition unnecessarily.

Reply:
So it looks like the best option is to put it in a different place under `pkg/agent`. Even if we do end up having a similar type in the API, it is confusing to use that type (internal or versioned) for the internal Agent implementation.

Reply:
Done